src/runtime/virtcontainers/nydus.go

```diff
@@ -0,0 +1,350 @@
+		nd.Logger().WithError(err).WithField("path", nd.sockPath).Warn("removing nydusd socket failed")
+	}
+	return nil
+}
+
+func (nd *nydusd) Logger() *log.Entry {
+	return virtLog.WithField("subsystem", "nydusd")
+}
+
+func (nd *nydusd) kill(ctx context.Context) (err error) {
+	span, _ := katatrace.Trace(ctx, nd.Logger(), "kill", virtiofsdTracingTags)
+	defer span.End()
+
+	if nd.pid == 0 {
+		return errors.New("invalid virtiofsd PID(0)")
+	}
+
+	err = syscall.Kill(nd.pid, syscall.SIGKILL)
+	if err != nil {
+		nd.pid = 0
+	}
+
+	return err
+}
+
+type BuildTimeInfo struct {
+	PackageVer string `json:"package_ver"`
+	GitCommit  string `json:"git_commit"`
+	BuildTime  string `json:"build_time"`
+	Profile    string `json:"profile"`
+	Rustc      string `json:"rustc"`
+}
+
+type DaemonInfo struct {
+	ID      string        `json:"id"`
+	Version BuildTimeInfo `json:"version"`
+	State   string        `json:"state"`
+}
+
+type ErrorMessage struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+}
+
+type MountOption struct {
+	mountpoint   string
+	bootstrap    string
+	daemonConfig string
+}
+type MountRequest struct {
+	FsType string `json:"fs_type"`
+	Source string `json:"source"`
+	Config string `json:"config"`
+}
+
+func NewMountRequest(fsType, source, config string) *MountRequest {
+	return &MountRequest{
+		FsType: fsType,
+		Source: source,
+		Config: config,
+	}
+}
+
+type Interface interface {
+	CheckStatus() (DaemonInfo, error)
+	Mount(string, *MountRequest) error
+	Umount(sharedMountPoint string) error
+}
+
+type NydusClient struct {
+	httpClient *http.Client
+}
+
+func NewNydusClient(sock string) (Interface, error) {
+	transport, err := buildTransport(sock)
+	if err != nil {
+		return nil, err
+	}
+	return &NydusClient{
+		httpClient: &http.Client{
+			Timeout:   defaultHttpClientTimeout,
+			Transport: transport,
+		},
+	}, nil
+}
+
+func waitUntilSocketReady(sock string) error {
+	b := &backoff.ExponentialBackOff{
+		InitialInterval:     300 * time.Microsecond,
+		RandomizationFactor: 0.5,
+		Multiplier:          2,
+		MaxInterval:         1 * time.Second,
+		MaxElapsedTime:      5 * time.Second,
+		Clock:               backoff.SystemClock,
+	}
+	b.Reset()
+	err := backoff.Retry(func() error {
+		_, err := os.Stat(sock)
+		return err
+	}, b)
+	if err != nil {
+		return errors.Wrap(err, "nydusd daemon socket not ready")
+	}
+	return nil
+}
+
+func buildTransport(sock string) (http.RoundTripper, error) {
+	err := waitUntilSocketReady(sock)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Transport{
+		MaxIdleConns:          10,
+		IdleConnTimeout:       10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+			dialer := &net.Dialer{
+				Timeout:   5 * time.Second,
+				KeepAlive: 5 * time.Second,
+			}
+			return dialer.DialContext(ctx, "unix", sock)
+		},
+	}, nil
+}
+
+func (c *NydusClient) CheckStatus() (DaemonInfo, error) {
```

Comment:
Is it possible we reuse/import corresponding package from nydus-snapshotter module?
They look quite similar.

Reply:
I'd prefer to avoid such extra dependency as the function is just a few (<20) lines.

Reply:
not only this function, NydusClient and the other struct almost come from nydus-snapshotter https://github.com/dragonflyoss/image-service/blob/master/contrib/nydus-snapshotter/pkg/nydussdk/client.go , I also think about import the package from nydus-snapshotter, but I also do not want add extra dependency.